engine/src/main/java/org/hibernate/search/engine/mapper/mapping/building/impl/IndexSchemaFilter.java

```diff
@@ -196,24 +206,91 @@ private String getPathFromSameIndexedEmbeddedSinceNoCompositionLimits(IndexedEmb
+	 * Let's assume we have a cycle where we are trying to add next `a.` definition same as the already added ones before:
+	 * a.b.c.a.b.c.a.b.c.a.b.c. + a.
+	 * 1   ^ 2     3     4        5
+	 *
+	 * - we'll be looking for a definition `a.` with index `1` and that would define our "longest cycle".
+	 * - we want to know the filter representing the definition `c.` (marked with `^`) that is right before the `a.` with index `2` as well as the
+	 * cycle-path leading to it from `5` -- "a.b.c.a.b.c.a.b.c."
+	 * that will be the filter and the path we'll pass to the "isPotentiallyExcludedPath" check. These will be our `previousCycleFilter` and `previousCycle`
+	 *
+	 * Doing so will go through any potential exclude filters defined on indexed-embedded only once; going from that definition `c.` (^)
+	 * and through its parents will cover the definitions from the cycle itself as well as any definitions out of the cycle that are in the path leading to the cycle.
+	 *
+	 * In case we don't have a repeating cycle yet - we'll be starting from the "current filter"
+	 * (previousCycleFilter == this, previousCycle == this.definition.relativePrefix() )
+	 */
+	private String findCycle(IndexedEmbeddedDefinition definitionToBeAdded) {
+		String shortestCycle = null;
+		String currentCycle = null;
+		String previousCycle = this.definition.relativePrefix();
+		IndexSchemaFilter localParent = this;
+		IndexSchemaFilter currentCycleFilter = null;
+		IndexSchemaFilter previousCycleFilter = this;
+		StringBuilder path = new StringBuilder();
+		while ( localParent != null && localParent.definition != null ) {
+
+			path.insert( 0, localParent.definition.relativePrefix() );
+
+			if ( localParent.definition.equals( definitionToBeAdded ) ) {
+
+				if ( currentCycleFilter != null ) {
+					previousCycleFilter = currentCycleFilter.parent;
+					previousCycle = currentCycleFilter.definition.relativePrefix() + currentCycle;
+				}
+
+				currentCycle = path.toString();
+				currentCycleFilter = localParent;
+
+				if ( shortestCycle == null ) {
+					// Set the first found cycle and keep it unchanged till the end of the loop. We'll report it to the user.
+					shortestCycle = currentCycle;
+				}
 			}
+
+			localParent = localParent.parent;
+		}
+
+		if ( currentCycle == null ) {
+			return null;
 		}
-		else {
-			/*
-			 * No recursion limits, no parent: this is the root.
-			 * I we reach this point, it means there was no recursion limit at all,
-			 * but we did not encounter the IndexedEmbedded we were looking for.
-			 */
+
+		// If we reached this point -- we've found a cycle, and we need to check if any exclude paths can break it:
+		if ( isPotentiallyExcludedPath(
+				// when we check the node we don't need the node's relative path at the beginning of the path we are checking:
+				previousCycle.substring( previousCycleFilter.definition.relativePrefix().length() ),
```

Comment:
Looks to me like you should rather adjust your loop so that `previousCycle` matches what you need.